### DISCUSSION TITLE
Better docusaurus package upgrade workflow

### DISCUSSION THREAD

**Discussion by HonkingGoose** [renovatebot/renovate#10948] [Request Help] [answered]
## Current situation

I filter out `canary` releases from Docusaurus with some `allowedVersions` regex.

Sometimes I do want the new stuff right away. To get the new stuff I need to manually change my `package.json` to point to the latest `canary` release, delete the lockfile and install the newer version(s) with npm.

After I manually switch to latest `canary` Renovate will not propose updates to any `allowedVersion` matching updates. So I have to do work twice, once to upgrade to canary, and once to upgrade to stable.

## Ideal situation

Ideally, I'd like a workflow like this:

1. Use stable releases by default, so updates that match the RegEx in `allowedVersions`.
2. Checkbox on Dependency Dashboard to get latest `canary` release.
3. Once Docusaurus releases a version that matches `allowedVersions` (stable release!), Renovate should update me to that.

## Discussion

I have no idea if something like this is even possible with the current Renovate code, or with the way that Docusaurus releases their packages.

## Current Renovate config

<details><summary>Current Renovate config</summary>

```json
{
  "extends": ["config:base", "helpers:pinGitHubActionDigests"],
  "dependencyDashboard": true,
  "prCreation": "not-pending",
  "internalChecksFilter": "strict",
  "lockFileMaintenance": {
    "enabled": true,
    "automerge": true
  },
  "packageRules": [
    {
      "description": "Wait 3 days before creating a npm update PR",
      "matchDatasources": ["npm"],
      "stabilityDays": 3
    },
    {
      "description": "Automerge mainline @docusaurus releases, ignore canary releases",
      "matchPackagePatterns": ["^@docusaurus"],
      "allowedVersions": "/^\\d+\\.\\d+\\.\\d+(-beta\\.(\\d+))?$/",
      "automerge": true
    },
    {
      "description": "Automerge Prettier updates",
      "matchDepTypes": ["devDependencies"],
      "matchPackagePatterns": ["prettier"],
      "automerge": true
    },
    {
      "description": "Automerge minor/patch React monorepo updates",
      "extends": "monorepo:react",
      "groupName": "react monorepo",
      "matchUpdateTypes": ["minor", "patch"],
      "automerge": true
    }
  ],
  "node": {
    "supportPolicy": ["lts_latest"]
  }
}

```

</details>

## Link to repository

https://github.com/HonkingGoose/git-gosling

**Comment by rarkins** [comment]
Normally Renovate will not upgrade from stable -> unstable. Isn't it the case that you're using beta releases, not stable ones? (according to the semver spec I mean)

**Reply by HonkingGoose** [reply]
This is a snippet from my `package.json` right now:

```json
  "dependencies": {
    "@docusaurus/core": "2.0.0-beta.3",
    "@docusaurus/preset-classic": "2.0.0-beta.3",
  },
```

Links to npm registry (check the version tab):

- https://www.npmjs.com/package/@docusaurus/core
- https://www.npmjs.com/package/@docusaurus/preset-classic

---

> Isn't it the case that you're using beta releases, not stable ones? (according to the semver spec I mean)

I did find this in the SemVer spec: https://semver.org/#spec-item-11

> Precedence for two pre-release versions with the same major, minor, and patch version MUST be determined by comparing each dot separated identifier from left to right until a difference is found as follows:
> 
>  1.   Identifiers consisting of only digits are compared numerically.
> 
>  2.   Identifiers with letters or hyphens are compared lexically in ASCII sort order.
> 
>   3.  Numeric identifiers always have lower precedence than non-numeric identifiers.
> 
>   4.  A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal.
> 
> Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.

Docusaurus also has `2.0.0-alpha.75` releases.... So I think `2.0.0-alpha` is considered newer than `2.0.0-beta` because it sorts in ASCII order?

---

Maybe I should just open a feature request over on Docusaurus if they could pretty please start using normal SemVer???

---

I guess the only options are:

1. Use latest canary
2. Use "stable" releases as filtered out by the `allowedVersion` RegEx

**Reply by rarkins** [reply]
> This is a snippet from my `package.json` right now:
> 
> ```json
>   "dependencies": {
>     "@docusaurus/core": "2.0.0-beta.3",
>     "@docusaurus/preset-classic": "2.0.0-beta.3",
>   },
> ```

This means you're using an *unstable* (beta) version, not stable one.

> > Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.
> 
> Docusaurus also has `2.0.0-alpha.75` releases.... So I think `2.0.0-alpha` is considered newer than `2.0.0-beta` because it sorts in ASCII order?

No, Semver sorts in the same order as alphabet, so beta is newer than alpha. It's just lucky/coincidental that alpha sorts before beta which sorts before rc.

What this means is that if you switch to an alpha canary release then Renovate calculates that the latest beta is newer and upgrades you.

> Maybe I should just open a feature request over on Docusaurus if they could pretty please start using normal SemVer???
> 
> I guess the only options are:
> 
> 1. Use latest canary
> 2. Use "stable" releases as filtered out by the `allowedVersion` RegEx

The challenge you have is that you want both of these:
- To switch between beta and canary at times
- For an automated bot to "follow your lead"

The docusaurus maintainers could make life easier in general using npm's dist-tags, although probably not to the point where you can easily switch between tags *and* have Renovate know what you're doing.

Right now they have these:
<img width="308" alt="image" src="https://user-images.githubusercontent.com/6311784/127285991-f6a1fd9c-ccea-4bac-a678-4f944ad0b5ec.png">

In theory they should have latest, next and canary pointing to stable, beta and canary respectively. Renovate could be used to follow such tags (`followTag`) feature but can't automatically switch that itself.

**Reply by HonkingGoose** [reply]
> The docusaurus maintainers could make life easier in general using npm's dist-tags, although probably not to the point where you can easily switch between tags _and_ have Renovate know what you're doing.
> 
> Right now they have these:
> <img alt="image" width="308" src="https://user-images.githubusercontent.com/6311784/127285991-f6a1fd9c-ccea-4bac-a678-4f944ad0b5ec.png">
> 
> In theory they should have latest, next and canary pointing to stable, beta and canary respectively. Renovate could be used to follow such tags (`followTag`) feature but can't automatically switch that itself.

Checking to see if I understand your idea, the `dist-tags` should be like this:

```markdown
dist-tags:
  latest: "1.14.7"
  next: "2.0.0-beta.3"
  canary: "2.0.0-beta.a7618c775"
```

How much work would it be for them to make this change? I'm not sure if it's reasonable for me to ask for things like this???

**Reply by rarkins** [reply]
Yes, that's how I'd recommend doing it. Without this, the semver spec has no real way to differentiate between the "semi-stable" beta versions and the canary ones.
The amount of work depends on whether they have an automated system for publishing or not..

**Reply by HonkingGoose** [reply]
> Yes, that's how I'd recommend doing it. Without this, the semver spec has no real way to differentiate between the "semi-stable" beta versions and the canary ones.

Okay so I got my example right. 😄 

> The amount of work depends on whether they have an automated system for publishing or not..

They don't use something like `release-please` or `semantic-release`. I think publishing is a semi-automated process, where a human is still needed for some steps.

The have a document describing their publishing workflow: https://github.com/facebook/docusaurus/blob/master/admin/publish.md

**Reply by slorber** [reply]
Hey, FYI v1 and v2 have different npm packages, and v1 had some legacy bad npm dist tags.
Also v2 packages only have beta  and canary, no official release yet but betas are published with latest dist tag.

After the cleanup:

![image](https://user-images.githubusercontent.com/749374/128181988-499f03dc-9e0a-40d2-9884-ea4760f2dd54.png)

Not sure how to improve but I feel like our canary version name is not ideal, it's not the first time I see issues with it (some people use a semver range with a canary hash in package.json and the npm/yarn ends up downloading another older canary version for example).

Do you think it would be better to have something like: `2.0.0-beta.4-8e9b829d9` ?

**Reply by HonkingGoose** [reply]
> Not sure how to improve but I feel like our canary version name is not ideal, it's not the first time I see issues with it (some people use a semver range with a canary hash in package.json and the npm/yarn ends up downloading another older canary version for example).

How about dropping the `canary` thing completely, and following the [SemVer 2.0 specification](https://semver.org/) instead? I see you already use some kind of conventional commit on the `facebook/docusaurus` repository.

I really like the way Renovate bot themselves release new versions. Each commit to `main` with a "release triggering" commit gets released automatically. Commits with the `chore:` and `docs:` prefix are bundled with the _next_ release triggering commit.
Commits with the `fix:` and `feat:` prefixes always trigger a new release. We manually release breaking changes in a new branch, which we then rebase merge into `main` when ready to go, example: https://github.com/renovatebot/renovate/pull/10995.

We use the [semantic-release](https://github.com/semantic-release/semantic-release) action to automate the release process.

If you wanted to keep manual control over _when_ to release a new "stable" version, you can take a look at the [release-please action](https://github.com/googleapis/release-please).

**Reply by slorber** [reply]
>. How about dropping the canary thing completely, and following the SemVer 2.0 specification instead? I see you already use some kind of conventional commit on the facebook/docusaurus repository.

That means moving out of beta, which would basically "cancel" what we committed to work on that beta phase: https://docusaurus.io/blog/2021/05/12/announcing-docusaurus-two-beta#what-are-the-goals-of-the-beta

Not sure we want to do that now. But in the future we could transform the canary channel to continuous delivery, similarly to what Renovate is doing, but it probably involve some work to find the right workflow for us.

**Reply by HonkingGoose** [reply]
> That means moving out of beta, which would basically "cancel" what we committed to work on that beta phase: https://docusaurus.io/blog/2021/05/12/announcing-docusaurus-two-beta#what-are-the-goals-of-the-beta

Oh okay, so that's a non-starter for now. 😄 

> But in the future we could transform the canary channel to continuous delivery, similarly to what Renovate is doing, but it probably involve some work to find the right workflow for us.

I think you said in another comment to me, that `canary` should be good to use in production as well? So that's why I proposed ditching `canary` in full and just say: if it's on `main` you can use it in production... 😄 But I'm probably missing something important, or else you would already follow that pattern. 😉

**Reply by slorber** [reply]
We still do releases with changelogs, twitter/discord announcements and probably more blog posts in the future so even if we have continuous releases in a quite robust way, we still want to mark some releases specifically as "latest" and have a good opportunity to market them correctly, and mention important features and potential breaking changes is a simpler way for users. Continuous delivery moves all that burden a bit to our users that would have to inspect commit histories...

**Reply by HonkingGoose** [reply]
Martin Fowler has a great article on different patterns to release code to production: https://martinfowler.com/articles/branching-patterns.html#release-ready-mainline

I think the Release-ready mainline pattern is closest to what you want?

> Release-ready mainline
>
> Keep mainline sufficiently healthy that the head of mainline can always be put directly into production

The release process could be automated with `release-please` which creates a PR of all commits that happened between the latest tagged version and `HEAD`. When you're ready to release, use `release-please` to bump the version number and generate the changelog. You can then make a big marketing push with a blog, socials, for the new version.

I recommend you take a look through the different patterns, maybe you see one that's a better fit for you and your team.

**Reply by rarkins** [reply]
Instead of publishing sha-based releases to npmjs, are canary users able to just depend directly on the source repo, or does it require "building" before it can be used? e.g. here's an example of depending directly on github repos from npm docs:

```
{
  "name": "foo",
  "version": "0.0.0",
  "dependencies": {
    "express": "expressjs/express",
    "mocha": "mochajs/mocha#4727d357ea",
    "module": "user/repo#feature\/branch"
  }
}
```

**Reply by slorber** [reply]
Sources are TS and we need to transpile it for both node and browser usage with different configs so it's not possible to depend on the repo directly.

@HonkingGoose this strategy looks fine but it's not something I'm planning to invest my time in right now. Maybe after the official release.

BTW to fix my issues once for all I've modified again the canary version name: https://github.com/facebook/docusaurus/pull/5295

Now using `0.0.0-3892` where `3892` is the commit number on master

**Reply by HonkingGoose** [reply]
> Sources are TS and we need to transpile it for both node and browser usage with different configs so it's not possible to depend on the repo directly.

Oh that's a shame, would have been even easier to make releases then... 😄 I'm not sure if `release-please` supports releasing to npm.js though, so check before you commit to any release helper bot/action/setup...

> @HonkingGoose this strategy looks fine but it's not something I'm planning to invest my time in right now. Maybe after the official release.

That's okay! I just wanted to mention the resource, and make you aware that _maybe_ there's a better/easier way to handle releases for your project.

> BTW to fix my issues once for all I've modified again the canary version name: [facebook/docusaurus#5295](https://github.com/facebook/docusaurus/pull/5295)
> 
> Now using `0.0.0-3892` where `3892` is the commit number on master

Cool! I'll go try this out on my repository, and see how Renovate handles the `canary` releases now. 😄

Thanks again for responding to my questions/ideas about versioning for Docusaurus, I'll stop bothering you about the versioning now. 😉

**Reply by slorber** [reply]
Our release process is a bit more complicated:
- generate changelog + add it manually to CHANGELOG.md
- create new docs version
- archive older docs versions, but keep dropdown link to older immutable Netlify deployment
- compile
- publish to npm
- create github release
- market on Discord/Twitter

It's probably a bit of work to find a way to automate all this without any manual intervention 😆 we'll figure this out later it's even worth it.

**Reply by viceice** [reply]
Changelog, GitHub release, compile and npm publish can be easy done with semantic release. We use it ourselves. 🤗

**Reply by travi** [reply]
also, the exec plugin can be used to run scripts as part of the release process for steps that don't have existing plugins: https://github.com/semantic-release/exec

**Comment by HonkingGoose** [comment, marked as answer]
I've opened an issue in the Docusaurus repository to discuss the use of the npm dist-tags:

- https://github.com/facebook/docusaurus/issues/5274

---

The `canary` versioning should now follow a incrementing versioning scheme, so it should now sort according to SemVer on the `canary` tag: 

- https://github.com/facebook/docusaurus/pull/5285

This means that I _should_ be able to use the Renovate `followTag` set to `canary` to get newer unstable releases from Renovate bot.
